pkg/provider/remote/remote.go

```diff
@@ -0,0 +1,265 @@
+// Copyright Nitric Pty Ltd.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/go-connections/nat"
+	"github.com/pkg/errors"
+	"github.com/pterm/pterm"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/nitrictech/cli/pkg/containerengine"
+	"github.com/nitrictech/cli/pkg/output"
+	"github.com/nitrictech/cli/pkg/provider/types"
+	deploy "github.com/nitrictech/nitric/core/pkg/api/nitric/deploy/v1"
+)
+
+type remoteDeployment struct {
+	cfc       types.ConfigFromCode
+	stackName string
+	provider  string
+	conn      *grpc.ClientConn
+	ce        containerengine.ContainerEngine
+	// Container id populated after a call to Start
+	cid string
+}
```

Comment:
Composability will help here for supporting more provider execution types:

```go
type binaryRemoteDeployment struct {
  *remoteDeployment
}

func (p *binaryRemoteDeployment) Up(log output.Progress) (*types.Deployment, error) {
	// TODO: Start binary provider ...	
        dep, err := p.remoteDeployment.Up(log)
	// TODO: Stop Binary provider

        return dep, nil
}
```

Would just need to remove the `conn` from the remote deployment struct, and create/close connections on a per use bases (move these to up/down with a share utility method for opening the connection.